reports/glides_validator.livemd

```diff
@@ -22,90 +24,553 @@ Mix.install(
-    local_path = remote_path |> Path.split() |> Enum.take(-1)
-    IO.inspect("Downloading #{remote_path} to #{local_path}...")
+    local_files = File.ls!(local_dir(env_suffix))
 
-    ExAws.S3.download_file(
-      "mbta-gtfs-s3",
-      remote_path,
-      local_path
+    tr_upd_file_counts =
+      populate_table(:TripUpdates, path_prefixes, env_suffix, local_files, sample_count)
+
+    ve_pos_file_counts =
+      populate_table(:VehiclePositions, path_prefixes, env_suffix, local_files, sample_count)
+
+    %{local: 0, downloaded: 0}
+    |> Map.merge(tr_upd_file_counts)
+    |> Map.merge(ve_pos_file_counts, fn _k, count1, count2 -> count1 + count2 end)
+  end
+
+  # Loads data into a table.
+  # Returns the number of files that were found locally,
+  # and the number that were newly downloaded.
+  defp populate_table(table_name, path_prefixes, env_suffix, local_files, sample_count) do
+    path_prefixes
+    |> Stream.map(&fetch_local_paths(&1, table_name, local_files, sample_count))
+    |> Task.async_stream(&do_load(&1, table_name, env_suffix, sample_count),
+      ordered: false,
+      timeout: 10000
     )
-    |> ExAws.request()
+    |> Stream.map(fn {:ok, result} -> result end)
+    |> Enum.group_by(fn {type, _} -> type end, fn {_, count} -> count end)
+    |> Map.new(fn {type, counts} -> {type, Enum.sum(counts)} end)
+  end
+
+  defp do_load({:ok, filenames}, table_name, env_suffix, _) do
+    dir = local_dir(env_suffix)
+    local_paths = Enum.map(filenames, &Path.join(dir, &1))
+    Enum.each(local_paths, &load_file_into_table(table_name, &1))
+    {:local, Enum.count(local_paths)}
+  end
+
+  defp do_load({:error, path_prefix, existing_matches}, table_name, env_suffix, sample_count) do
+    local_paths =
+      download_files(path_prefix, table_name, env_suffix, sample_count, existing_matches)
+
+    Enum.each(local_paths, &load_file_into_table(table_name, &1))
+    {:downloaded, Enum.count(local_paths)}
+  end
+
+  defp local_dir(""), do: Path.join([System.tmp_dir!(), "glides_report", "prod"])
+  defp local_dir("-" <> env), do: Path.join([System.tmp_dir!(), "glides_report", env])
+
+  defp fetch_local_paths(path_prefix, table_name, local_files, sample_count) do
+    file_prefix = Path.basename(path_prefix)
+    file_substring = "https_cdn.mbta.com_realtime_#{table_name}_enhanced.json"
+
+    matches =
+      Enum.filter(local_files, fn filename ->
+        String.starts_with?(filename, file_prefix) and
+          String.contains?(filename, file_substring)
+      end)
+
+    if length(matches) >= sample_count do
+      {:ok, Enum.take(matches, sample_count)}
+    else
+      {:error, path_prefix, MapSet.new(matches)}
+    end
+  end
+
+  # Downloads VehiclePosition or TripUpdate files and returns the local paths they were downloaded to.
+  defp download_files(remote_prefix, table_name, env_suffix, sample_count, existing_matches) do
+    IO.inspect("mbta-gtfs-s3#{env_suffix}", label: "bucket name")
```

Comment:
Are we communicating which bucket is being queried well enough?
![Screenshot 2024-06-26 at 9 56 10 AM](https://github.com/mbta/transit_data_reports/assets/304505/879fc14c-c4a8-4536-a470-430ed152be1e)

Reply:
Whoops! Thank you, @Whoops